 
 3. URL for zoneedit changed to http://dynamic.zoneedit.com
 4. HTTP/301 redirect support by theMIROn <themiron@mail.ru>
 5. Format string vulnerability in syslog handling
    http://bugs.gentoo.org/show_bug.cgi?id=69658
 7. changed maxintervals down to 1 hour
 
diff -BurpN -x'*.o' router/ez-ipupdate/ez-ipupdate.c gateway/ez-ipupdate/ez-ipupdate.c
--- router/ez-ipupdate/ez-ipupdate.c	2004-11-11 07:26:51.000000000 +0000
+++ gateway/ez-ipupdate/ez-ipupdate.c	2009-06-27 05:00:40.000000000 +0000
@@ -44,6 +44,9 @@
 #define SEND_EMAIL_CMD "mail"
 #endif
 
+#undef HAS_REDIRECT
+// support of HTTP/301 - redirect command
+
 #define EZIP_DEFAULT_SERVER "www.EZ-IP.Net"
 #define EZIP_DEFAULT_PORT "80"
 #define EZIP_REQUEST "/members/update/"
@@ -102,7 +105,7 @@
 #define HN_DEFAULT_PORT "80"
 #define HN_REQUEST "/vanity/update"
 
-#define ZONEEDIT_DEFAULT_SERVER "www.zoneedit.com"
+#define ZONEEDIT_DEFAULT_SERVER "dynamic.zoneedit.com"
 #define ZONEEDIT_DEFAULT_PORT "80"
 #define ZONEEDIT_REQUEST "/auth/dynamic.html"
 
@@ -217,11 +224,11 @@
 #define MIN_UPDATE_PERIOD 10
 // the min/max time to wait if we fail to update
 #define MIN_WAIT_PERIOD 300
-#define MAX_WAIT_PERIOD (2*3600)
+#define MAX_WAIT_PERIOD (1*3600)
 // the min time that max-period can be set to
-#define MIN_MAXINTERVAL (24*3600)
+#define MIN_MAXINTERVAL (1*3600)
 // the max time we will wait if the server tells us to
-#define MAX_WAITRESPONSE_WAIT (24*3600)
+#define MAX_WAITRESPONSE_WAIT (1*3600)
 #define MAX_MESSAGE_LEN 256
 #define ARGLENGTH 32
 #define BLOCK_FILE "/var/run/ez-ipupdate.block"
@@ -859,7 +876,7 @@ void show_message(char *fmt, ...)
     sprintf(buf, "message incomplete because your OS sucks: %s\n", fmt);
 #endif
 
-    syslog(LOG_NOTICE, buf);
+    syslog(LOG_NOTICE, "%s", buf);
   }
   else
   {
@@ -885,7 +902,7 @@ void show_message(char *fmt, ...)
   va_start(args, fmt);
   vsnprintf(buf, sizeof(buf), fmt, args);
   openlog("ddns update", 0, 0);
-  syslog(0, buf);
+  syslog(0, "%s", buf);
   closelog();
   va_end(args);
 }
@@ -4057,19 +4074,28 @@ int ZONEEDIT_update_entry(void)
   int bytes;
   int btot;
   int ret;
+#ifdef HAS_REDIRECT
+  char bserver[128];
+  char brequest[128];
+  int redir_count = 0;
+#endif
+  char *lserver = server;
+  char *lrequest = request;
+
+  redirect_loop:
 
   buf[BUFFER_SIZE] = '\0';
 
-  if(do_connect((int*)&client_sockfd, server, port) != 0)
+  if(do_connect((int*)&client_sockfd, lserver, port) != 0)
   {
     if(!(options & OPT_QUIET))
     {
-      show_message("error connecting to %s:%s\n", server, port);
+      show_message("error connecting to %s:%s\n", lserver, port);
     }
     return(UPDATERES_ERROR);
   }
 
-  snprintf(buf, BUFFER_SIZE, "GET %s?", request);
+  snprintf(buf, BUFFER_SIZE, "GET %s?", lrequest);
   output(buf);
   snprintf(buf, BUFFER_SIZE, "%s=%s&", "host", host);
   output(buf);
@@ -4086,7 +4112,7 @@ int ZONEEDIT_update_entry(void)
   snprintf(buf, BUFFER_SIZE, "User-Agent: %s-%s %s (%s)\015\012", 
       "zoneedit", VERSION, OS, "by Angus Mackay");
   output(buf);
-  snprintf(buf, BUFFER_SIZE, "Host: %s\015\012", server);
+  snprintf(buf, BUFFER_SIZE, "Host: %s\015\012", lserver);
   output(buf);
   snprintf(buf, BUFFER_SIZE, "Authorization: Basic %s\015\012", auth);
   output(buf);
@@ -4141,6 +4167,35 @@ int ZONEEDIT_update_entry(void)
       }
       break;
 
+#ifdef HAS_REDIRECT
+    case 301:
+      *bserver = '\0';
+      *brequest = '\0';
+      if(((bp=strstr(buf, "Location: ")) != NULL) &&      
+	 ((ret=sscanf(bp,"Location: http://%127[^/\r\n]%127[^?\r\n]", bserver, brequest)) > 0) &&
+	 (redir_count++ <= 5))
+      {
+	if(ret==1) snprintf(brequest, 127, "/");
+	lserver = bserver;
+	lrequest = brequest;
+	if(!(options & OPT_QUIET))
+	{
+	  show_message("trying location %u: %s%s\n", redir_count, lserver, lrequest);
+	}
+	goto redirect_loop;
+      }
+      else
+      {	
+	show_message("error processing redirect\n");
+	if(!(options & OPT_QUIET))
+	{
+	  fprintf(stderr, "server output: %s\n", buf);
+        }
+      }
+      return(UPDATERES_ERROR);
+      break;
+#endif
+
     case 401:
       if(!(options & OPT_QUIET))
       {
